### PR TITLE
Riscv port

### PR DESCRIPTION
**Pull request by ct-clmsn**
### Ticket

Link to Github Issue #14909 

### Problem description

Provides preliminary support for rv64g sifive unmatched hardware.

This patch generalizes several existing x86 specific portions of the tt-metal stack such that they are now portable across multiple architectures - the x86 intrinsics have been ported to ISO C++.

Part of this port modifies the existing build script so that users can now select different C and C++ compilers (and versions of compilers) to build tt-metal.  The selected compilers properly forward into to the underlying tt-metal CMake infrastructure.

Thanks to @joelsmithTT for making time for discussion and verification.

### What's changed

Small portions of the existing C++ functionality use x86 specific intrinsics. This patch modifies those portions in order to generalize them to ISO C++. This work relied heavily on the following [link](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html) which describes in general terms (pseudo code) what operations a particular set of x86 intrinsics perform.

The bash script that runs the build has been updated to allow users to select compilers (C/C++).

A substantial number of unit tests pass but not all of them. There are several instances where a unit test composed of several other unit tests pass but, when the dependency unit tests are executed  they fail with a segmentation fault.

Some tests also seem to require `TT_METAL_SLOW_DISPATCH_MODE=1` but do not state it's a requirement unless they are run independently. As an example, running the "unit_test" program will run several unit tests without that environment variable activated. Several of the tests run as part of "unit_test" will fail. Some of tests that fail can run correctly with that environment variable outside of being run as part of the umbrella suite found in "unit_test".

### Checklist
- [ ] Post commit CI passes
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: build_metal.sh [hunk 1]
```diff
@@ -155,10 +168,18 @@ echo "INFO: Build tests: $build_tests"
 
 # Prepare cmake arguments
 cmake_args+=("-B" "$build_dir")
-cmake_args+=("-G" "Ninja")
```
Comment (blozano-tt) [inline build_metal.sh:158]: Why remove Ninja?

File: scripts/build_scripts/build_with_profiler_opt.sh [hunk 2]
```diff
@@ -1,4 +1,9 @@
 #! /usr/bin/env bash
+#
+# Copyright
+# © 2023 Tenstorrent Inc.
+# © 2024 Tactical Computing Labs, LLC
+#
```
Comment (blozano-tt) [inline scripts/build_scripts/build_with_profiler_opt.sh:6]: This script was deleted from main. I don't understand why I am seeing it in this diff.

File: CMakeLists.txt [hunk 3]
```diff
@@ -191,7 +197,8 @@ ADJUST_COMPILER_WARNINGS()
 
 add_library(compiler_flags INTERFACE)
 target_link_libraries(compiler_flags INTERFACE linker_flags compiler_warnings stdlib)
-target_compile_options(compiler_flags INTERFACE -mavx2 -fPIC -DFMT_HEADER_ONLY -fvisibility-inlines-hidden -fno-lto)
+#target_compile_options(compiler_flags INTERFACE -mavx2 -fPIC -DFMT_HEADER_ONLY -fvisibility-inlines-hidden -fno-lto)
```
Comment (blozano-tt) [inline CMakeLists.txt:200]: Please cleanup dead code

**Comment by ayerofieiev-tt** [comment]
Hi @ct-clmsn! This is awesome! Thrilled to see this change coming in! 🤩


We don't have RISCV machines on CI yet. Can you please share what environment do you use to build and test this? 
This might help us to establish some CI.

File: build_metal.sh (hunk 1 above)
Comment (ct-clmsn) [reply build_metal.sh]: This is an artifact of my previous attempt to do a riscv build back in April. Ninja had some issues. I just added that line back into the script and everything is working find. Thanks for finding this one and asking.

File: scripts/build_scripts/build_with_profiler_opt.sh (hunk 2 above)
Comment (ct-clmsn) [reply scripts/build_scripts/build_with_profiler_opt.sh:6]: This patch was built on a specific commit that is out of sync with main; this commit point was suggested by some folks at TT. This particular script had references to Ninja.

File: CMakeLists.txt (hunk 3 above)
Comment (ct-clmsn) [reply CMakeLists.txt:200]: will do, thanks

File: scripts/build_scripts/build_with_profiler_opt.sh (hunk 2 above)
Comment (ct-clmsn) [reply scripts/build_scripts/build_with_profiler_opt.sh:6]: I've updated this branch to mirror main; will hopefully have this branch synchronized soon.

Comment (ct-clmsn) [reply scripts/build_scripts/build_with_profiler_opt.sh:6]: branch should be synchronized now. I've noticed the cmake scripts require `umd::device` before it's compiled by the build script. I've build umd independent of the build script and added in `find_package` to try to detect an existing build and installation of the dependency.

File: CMakeLists.txt (hunk 3 above)
Comment (ct-clmsn) [reply CMakeLists.txt]: should be removed now

File: build_metal.sh (hunk 1 above)
Comment (ct-clmsn) [reply build_metal.sh:158]: I've corrected this issue; added ninja back into the script. the PR should be synchronized now.

**Comment by ct-clmsn** [comment]
@blozano-tt apologies for the side effects of doing the rebase

File: tt_metal/common/bfloat8.hpp [hunk 4]
```diff
@@ -248,12 +284,122 @@ inline std::vector<float> unpack_bfp8_tiles_into_float_vec(const std::vector<uin
...
+                        std::vector<std::uint32_t> combined(8, 0);
+
+                        for(std::size_t j = 0; j < 8; ++j) {
+			   if(shift[j] < 32) {
+   			      first[j] = first[j] & mask[j] >> shift[j];
+			   }
+			   else {
+   			      first[j] = 0;
+			   }
+			}
```
Comment (joelsmithTT) [inline tt_metal/common/bfloat8.hpp:306]: It looks like this will perform out-of-bounds access on `first` (size == 4, loop iterates 8 times)

Comment (joelsmithTT) [reply tt_metal/common/bfloat8.hpp:306]: tests/tt_metal/tt_metal/test_bfp8_conversion.cpp has a test case for the conversion code in bfloat8.hpp

Comment (ct-clmsn) [reply tt_metal/common/bfloat8.hpp:306]: Nice catch, thanks for the code review. I'll make updates this evening.

**Comment by jvegaTT** [comment]
I am one of the code owners for the CCL files. We are seeing all the CCL changes from the past few months in the diff in this PR with 92 affected files just in the CCL side, it is impossible to filter out what was actually changed if anything in this PR pertaining to our area. I imagine there is a similar problem being faced by others as there are 2507 changed files! I don't think we can reasonably review this. I suggest reverting to an earlier commit before the erroneous merge with main and to perform a rebase instead.

**Comment by ct-clmsn** [comment]
@jvegaTT not a problem; completely understand the situation. Not exactly sure how the PR went sideways but I'll see about getting this fixed this evening. would you all prefer the branch get a hard reset to main or the branch get deleted and replaced with a new branch that's been created from main?

**Comment by SeanNijjar** [comment]
> @jvegaTT not a problem; completely understand the situation. Not exactly sure how the PR went sideways but I'll see about getting this fixed this evening. would you all prefer the branch get a hard reset to main or the branch get deleted and replaced with a new branch that's been created from main?

I _think_ GH will need the pr reopened to have all the extra code owners and changes removed from it.

**Comment by ct-clmsn** [comment]
Cool, I'll close this and reopen. Thanks.

**Pull request closed without merging**